torch_xla/csrc/runtime/distributed_runtime.h

```diff
@@ -3,33 +3,78 @@
 
 #include <memory>
 
+#include "torch_xla/csrc/runtime/debug_macros.h"
+#include "tsl/distributed_runtime/preemption/preemption_sync_manager.h"
 #include "xla/pjrt/distributed/distributed.h"
 
 namespace torch_xla {
 namespace runtime {
 
+// DistributedRuntime serves as the point of entry for all operations which
```

Comment:
Thanks for improving this class, Jon!

Reply:
Whew, thanks for the tough questions Xiongfei! Really pushing my C++ knowledge to the limits 😛

I've added @will-cromar to this review as well, since he's much more c++ savvy than I am. Will, could you please double check my understanding? Mainly around the singleton instance handling.

Reply:
I have one general question: why don't we tie the lifetime of `DistributedRuntime` to `ComputationClient` by having it hold a `unique_ptr<DistributedRuntime>`? Is it because we never run the destructor on `ComputationClient`? Or does the lifetime of `DistributedRuntime` substantially differ from that of the `PjRtClient`?

I ask because IMO one singleton is better than two.

(If the above change is possible but non-trivial, one of us can do the refactoring in a follow-up)

Reply:
Actually, we should do absolutely that because of the GPU case - we should never shut down the dist runtime without also shutting down the ComputationClient for GPUs.

This helped with testability for TPUs, since it's currently easy to recreate the dist runtime without killing the computation client, but I really just need to recreate the PreemptionSyncManager between tests.

Reply:
Refactored into ComputationClient